File: day5/day5.py
#!/usr/bin/env python3

def draw(m, start, end):
    x0, y0 = start
    x1, y1 = end
    sx = (x0 < x1) and 1 or (x0 > x1) and -1 or 0
    sy = (y0 < y1) and 1 or (y0 > y1) and -1 or 0

    while True:
        m[x0, y0] = m.get((x0, y0), 0) + 1
        if x0 == x1 and y0 == y1:
            break
        x0 += sx
        y0 += sy

def part1(segments):
    m = {}
    for s, e in segments:
        if s[0] == e[0] or s[1] == e[1]:
            draw(m, s, e)
    return sum(1 for x in m.values() if x > 1)

def part2(segments):
    m = {}
    for s, e in segments:
        draw(m, s, e)
    return sum(1 for x in m.values() if x > 1)

import sys

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read()

        segments = []
        for line in txt.splitlines():
            s0, s1 = line.split("->")
            s0 = tuple(int(x) for x in s0.split(","))
            s1 = tuple(int(x) for x in s1.split(","))
            segments.append((s0, s1))

        print("Part1:", part1(segments))
        print("Part2:", part2(segments))
    except FileNotFoundError:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)

File: day5/day5.cpp
#include <fstream>
#include <vector>
#include <unordered_map>

#include <fmt/format.h>

using namespace std;

struct Segment
{
	int x0, y0;
	int x1, y1;
};

istream& operator>>(istream& in, Segment& s)
{
	in >> s.x0;
	in.get();		// ,
	in >> s.y0;
	in.seekg(4, ios_base::cur); // " => "
	in >> s.x1;
	in.get();		// ,
	in >> s.y1;
	return in;
}

class SparseMap
{
public:
	void draw(int x0, int y0, int x1, int y1)
	{
		int sx = x0 < x1 ? 1 : x0 > x1 ? -1 : 0;
		int sy = y0 < y1 ? 1 : y0 > y1 ? -1 : 0;
		while (true)
		{
			map[pair<int,int>(x0, y0)]++;
			if (x0 == x1 && y0 == y1)
			{
				break;
			}
			x0 += sx;
			y0 += sy;
		}
	}

	size_t count() const
	{
		size_t count = 0;
		for (const auto& el : map)
		{
			if (el.second > 1)
			{
				count++;
			}
		}
		return count;
	}

	void reset()
	{
		map.clear();
	}

private:
	struct HashPoint
	{
		size_t operator()(const pair<int,int>& k) const
		{
			return k.first + (k.second<<8);
		}
	};
	unordered_map<pair<int,int>, size_t, HashPoint> map;
};

static size_t part1(SparseMap &map, const vector<Segment>& segs)
{
	map.reset();
	for (const auto& s : segs)
	{
		if (s.x0 == s.x1 || s.y0 == s.y1)
		{
			map.draw(s.x0, s.y0, s.x1, s.y1);
		}
	}
	return map.count();
}

static size_t part2(SparseMap &map, const vector<Segment>& segs)
{
	map.reset();
	for (const auto& s : segs)
	{
		map.draw(s.x0, s.y0, s.x1, s.y1);
	}
	return map.count();
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fmt::print(stderr, "Usage: {} <filename>\n", argv[0]);
		return 1;
	}

	ifstream in(argv[1]);
	if (!in)
	{
		fmt::print(stderr, "Cannot open {}\n", argv[1]);
		return 1;
	}

	vector<Segment> segs;
	Segment tmp;
	while (in >> tmp)
	{
		segs.push_back(move(tmp));
	}
	in.close();

	SparseMap map;
	fmt::print("Part1: {}\n", part1(map, segs));
	fmt::print("Part2: {}\n", part2(map, segs));
	return 0;
}
